# Predicting Average House Prices in Madrid Neighbourhoods Using Socioeconomic Indicators

## Overall Project Aim:
To develop and evaluate a machine learning model that uses 2025 socioeconomic indicators to predict average housing prices across Madrid neighbourhoods.

## 2. Feature Selection

### Introduction
Following the initial exploration of the raw data, the socioeconomic indicators included in the dataset will need to be examined to identify appropriate features for the machine learning model. The purpose of this notebook is to explore these indicators and select relevant features for subsequent analysis and modelling.

### Import Libraries

In [1]:
# Import the Pandas library for data manipulation and analysis
import pandas as pd

### Load Dataset

In [2]:
# Loading Dataset 1: Panel of Sociodemographic Indicators of Districts and Neighborhoods of Madrid
indicators = pd.read_csv("../data/indicators.csv",sep=";")

In [3]:
# Create a clean working copy of dataset
indicators_clean = indicators.copy()

### Initial Inspection and Column Rename

In [4]:
# Display the first five rows
indicators_clean.head()

,Orden,Periodo panel,ciudad,cod_distrito,distrito,cod_barrio,barrio,año,fecha_indicador,fuente_indicador,categoría_1,categoría_2,indicador_nivel1,indicador_nivel2,indicador_nivel3,unidad_indicador,indicador_completo,valor_indicador
0,1,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Superficie,NaN,NaN,Ha,Superficie (ha.),523
1,2,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Densidad de población,NaN,NaN,Habitantes/Ha.,Población densidad (hab./ha.),258
2,3,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,NaN,NaN,Habitantes,Número habitantes,134881
3,4,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,NaN,Habitantes,Población hombres,67829
4,5,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,67052


In [5]:
# Rename columns

indicators_clean.rename(columns={
    "Orden": "row_no",
    "Periodo panel": "dataset_version",
    "ciudad": "city",
    "cod_distrito": "district_code",
    "distrito": "district_name",
    "cod_barrio": "neighbourhood_code",
    "barrio": "neighbourhood_name",
    "año": "year",
    "fecha_indicador": "date",
    "fuente_indicador": "data_source",
    "categoría_1": "main_category",
    "categoría_2": "subcategory",
    "indicador_nivel1": "indicator_level1",
    "indicador_nivel2": "indicator_level2",
    "indicador_nivel3": "indicator_level3",
    "unidad_indicador": "unit",
    "indicador_completo": "indicator_name",
    "valor_indicador": "indicator_value"
}, inplace=True)


In [6]:
# Display the first ten rows following column rename
indicators_clean.head(10)

,row_no,dataset_version,city,district_code,district_name,neighbourhood_code,neighbourhood_name,year,date,data_source,main_category,subcategory,indicator_level1,indicator_level2,indicator_level3,unit,indicator_name,indicator_value
0,1,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Superficie,NaN,NaN,Ha,Superficie (ha.),523
1,2,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Densidad de población,NaN,NaN,Habitantes/Ha.,Población densidad (hab./ha.),258
2,3,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,NaN,NaN,Habitantes,Número habitantes,134881
3,4,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,NaN,Habitantes,Población hombres,67829
4,5,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,67052
5,6,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Edad Media,NaN,NaN,Edad,Edad media de la población,"44,03"
6,7,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,0-14,NaN,Personas,Población de 0 a 14 años,10554
7,8,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,15-29,NaN,Personas,Población de 15 a 29 años,21800
8,9,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,30-44,NaN,Personas,Población de 30 a 44 años,42857
9,10,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,45-64,NaN,Personas,Población de 45 a 64 años,38025


In [7]:
# Checking dataset coverage across different years
indicators_clean["dataset_version"].value_counts().sort_index()

dataset_version
2020    18846
2021    19493
2022    18906
2023    19578
2024    19569
2025    20516
Name: count, dtype: int64

### Observations 
As the model aims to use the most recent available data, 2025 was selected as the year of analysis. A check was conducted to confirm that 2025 had comparable data coverage to other recorded years, helping to validate its suitability for use in the model. 

### Inspecting Neighbhourhood Names

In [8]:
#Display unique dataset names
indicators_clean["neighbourhood_name"].nunique()

131

In [9]:
# Remove rows where neighbourhood_name is missing
indicators_clean_n= indicators_clean.dropna(subset=["neighbourhood_name"])

In [10]:
indicators_clean_n.sample(10)

,row_no,dataset_version,city,district_code,district_name,neighbourhood_code,neighbourhood_name,year,date,data_source,main_category,subcategory,indicator_level1,indicator_level2,indicator_level3,unit,indicator_name,indicator_value
8009,8010,2020,Ciudad de Madrid,10.0,Latina,105.0,Campamento,2019,01/02/2020,Estadística Ayuntamiento de Madrid,Indicadores Población desempleo,Población desempleo,Número de personas paradas,Hombres,NaN,Personas,Paro registrado (número de personas registrada...,388
94805,94806,2024,Ciudad de Madrid,20.0,San Blas-Canillejas,204.0,Arcos,2023,28/05/2023,Estadística Ayuntamiento de Madrid,Resultados Elecciones Locales,NaN,Número de votos,Censo,Municipio,Votos,"Elecciones municipales, censo electoral",18550
112075,112076,2025,Ciudad de Madrid,17.0,Villaverde,175.0,Ángeles,2023,01/01/2023,Estadística Ayuntamiento de Madrid,Población del distrito,Dinámica demográfica,Crecimiento Demográfico,NaN,NaN,Tasa,Tasa de crecimiento demográfico (porcentaje),"-5,490086568"
30163,30164,2021,Ciudad de Madrid,15.0,Ciudad Lineal,152.0,Pueblo Nuevo,2019,01/01/2019,Portal de datos abiertos del Ayuntamiento de M...,Vivienda,NaN,Tiempo,Media de crédito,NaN,Tiempo (meses),Duración media del crédito (meses) en transacc...,305
45403,45404,2022,Ciudad de Madrid,10.0,Latina,102.0,Puerta del Ángel,2022,01/02/2022,Estadística Ayuntamiento de Madrid,Indicadores Población desempleo,Indicadores Desempleo,Número de personas paradas,Mujeres,De 45 a 64 años,Tasa,Tasa de desempleo en mujeres de 45 a 64 años,"12,87"
35928,35929,2021,Ciudad de Madrid,12.0,Usera,126.0,Zofío,2019,26/05/2019,Estadística Ayuntamiento de Madrid,Resultados Elecciones Locales,NaN,Número de votos,Más Madrid,Municipio,Votos,"Elecciones municipales, votos a Más Madrid",1962
37748,37749,2021,Ciudad de Madrid,5.0,Chamartín,53.0,Ciudad Jardín,2021,01/01/2021,Portal de datos abiertos del Ayuntamiento de M...,Empresa y comercio,Locales en uso como vivienda,Locales,Uso vivienda,NaN,Número,Número de locales dados de alta en uso como vi...,16
100684,100685,2025,Ciudad de Madrid,5.0,Chamartín,56.0,Castilla,2023,28/05/2023,Estadística Ayuntamiento de Madrid,Resultados Elecciones Locales,NaN,Número de votos,Ciudadanos,Municipio,Votos,"Elecciones municipales, votos a Ciudadanos-Par...",323
82854,82855,2024,Ciudad de Madrid,8.0,Fuencarral-El Pardo,88.0,El Goloso,2023,01/01/2023,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Población mayor de 65 años/Población total,NaN,Proporción,Porcentaje de envejecimiento (población mayor ...,"8,51"
8280,8281,2020,Ciudad de Madrid,10.0,Latina,103.0,Lucero,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,Nacionalidad española,Personas,Personas con nacionalidad española hombres,13991


In [11]:
# Obtaining a sample of neighbourhood names to compare for obtaining a foreign key to later merging

indicators_clean_n[
    ["neighbourhood_code", "neighbourhood_name"]
].drop_duplicates().head(20)

,neighbourhood_code,neighbourhood_name
186,11.0,Palacio
269,12.0,Embajadores
351,13.0,Cortes
425,14.0,Justicia
508,15.0,Universidad
591,16.0,Sol
995,21.0,Imperial
1070,22.0,Acacias
1148,23.0,Chopera
1227,24.0,Legazpi


### Observations
Inspection of the neighbourhood names shows that accents are retained; however, unlike the housing prices dataset, the names do not contain numerical prefixes or punctuation (e.g '011. Palacio' vs 'Palacio) . Therefore, it is was decided to standardise the neighbourhood names in the housing prices dataset to match the format used in this dataset, creating a consistent key for merging.

### Exploring Neighbourhood-Level Indicators and Categories

In [12]:
# Display information for one neighbourhood in 2025

palacio = indicators_clean_n.loc[
    (indicators_clean_n['dataset_version'] == 2025) &
    (indicators_clean_n['neighbourhood_name']== 'Palacio')]
palacio


,row_no,dataset_version,city,district_code,district_name,neighbourhood_code,neighbourhood_name,year,date,data_source,main_category,subcategory,indicator_level1,indicator_level2,indicator_level3,unit,indicator_name,indicator_value
96528,96529,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Superficie,NaN,NaN,Ha,Superficie (ha.),"146,99"
96529,96530,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Densidad de población,NaN,NaN,Habitantes/Ha.,Población densidad (hab./ha.),"163,2832165"
96530,96531,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,NaN,NaN,Habitantes,Número habitantes,24001
96531,96532,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,NaN,Habitantes,Población hombres,11693
96532,96533,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,12308
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
116034,116035,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2023,01/01/2023,Estadística Ayuntamiento de Madrid,Indicadores Económicos,Indicadores económicos,Renta,Neta,Per cápita,Renta,Renta media disponible por persona,23770
116190,116191,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Portal de datos abiertos del Ayuntamiento de M...,Empresa y comercio,Total locales,Locales,Total,NaN,Número,Número de locales dados de alta totales,3062
116191,116192,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Portal de datos abiertos del Ayuntamiento de M...,Empresa y comercio,Locales abiertos,Locales,Abiertos,NaN,Número,Número de locales dados de alta abiertos,2617
116192,116193,2025,Ciudad de Madrid,1.0,Centro,11.0,Palacio,2025,01/01/2025,Portal de datos abiertos del Ayuntamiento de M...,Empresa y comercio,Locales cerrados,Locales,Cerrados,NaN,Número,Número de locales dados de alta cerrados,445


In [13]:
# Display the diffrerent main categories of the indicators
palacio["main_category"].nunique()

11

In [14]:
# Display the counts across the main categories of the indicators

palacio["main_category"].value_counts()

main_category
Población del distrito                           32
Educación                                        22
Servicios y equipamientos Municipales            16
Indicadores Población desempleo                  12
Resultados Elecciones Locales                    10
Vivienda                                          6
Índice de vulnerabilidad                          5
Empresa y comercio                                5
Participación ciudadana                           5
Indicadores Económicos                            4
Características Generales del Distrito-Barrio     2
Name: count, dtype: int64

English translation of main categories (Source: Google translate)

main_category
District population                              32
Education                                        22
Municipal services and facilities                16
Population and unemployment indicators           12
Local election results                           10
Housing                                           6
Vulnerability index                               5
Business and commerce                             5
Citizen participation                             5
Economic indicators                               4
General characteristics of the district/neighborhood 2
Name: count, dtype: int64

In [15]:
# Display the diffrerent subcategories of the indicators
palacio["subcategory"].nunique()

18

In [16]:
# Display the different subcategories of the indicators

palacio["subcategory"].value_counts()

subcategory
Estructura de población                 24
Población en etapas educativas          15
Población desempleo                     12
Servicios Sociales Municipales           8
Nivel de estudios de la población        7
Estructura de los hogares                6
Territorial                              5
Entidades de participación ciudadana     5
Deportes                                 4
Indicadores económicos                   4
Dinámica demográfica                     2
Cultura                                  2
Educación                                2
Recursos municipales                     1
Total locales                            1
Locales abiertos                         1
Locales cerrados                         1
Locales en uso como vivienda             1
Name: count, dtype: int64

English translation of subcategories (Source: Google translate)

subcategory
Population structure                   24
Population by educational stage        15
Unemployment                           12
Municipal social services               8
Population's educational attainment     7
Household structure                     6
Territorial                             5
Citizen participation bodies            5
Sports                                  4
Economic indicators                     4
Demographic dynamics                    2
Culture                                 2
Education                               2
Municipal resources                     1
Total premises                          1
Open premises                           1
Closed premises                         1
Premises used as housing                1
Name: count, dtype: int64

### Observations

Exploration of the 2025 data for the Palacio neighbourhood shows 119 rows of information, covering 11 main categories and 18 subcategories. While these categories can provide useful context for organising the data, observations are heavily concentrated within a small number of categories. Therefore, feature selection will focus primarily on the individual indicator_name values, which provide more specific information about the socioeconomic characteristics available for modelling.

In [17]:
# Check the number of indicator records per neighbourhood in 2025
neighbourhood_catalog = (indicators_clean_n[indicators_clean_n["dataset_version"] == 2025]
    .groupby("neighbourhood_name")
    .size()
    .reset_index(name="row_count")
)
neighbourhood_catalog

,neighbourhood_name,row_count
0,Abrantes,111
1,Acacias,115
2,Adelfas,115
3,Aeropuerto,107
4,Alameda de Osuna,116
...,...,...
126,Vinateros,112
127,Vista Alegre,119
128,Zofío,114
129,Águilas,119


In [18]:
# Saving the number of indicator records per neighbourhood in 2025 in an excel
neighbourhood_catalog.to_excel(
    "../outputs/neighbourhood_catalog.xlsx",
    index=False
)

### Observations 

The 131 neighbourhoods contain between 106 and 122 indicator records for 2025. While the number of records varies, overall indicator coverage appears broadly similar across neighbourhoods.

In [19]:
# Retain only columns with information on types of indicies using palacio 
n_indicator_catalog= palacio[
    [
        "main_category",
        "subcategory",
        "indicator_name",
        "unit"
    ]]

In [20]:
# Remove duplicates
n_indicator_catalog = n_indicator_catalog.drop_duplicates()


In [21]:
# Sort alphabetically
n_indicator_catalog = n_indicator_catalog.sort_values(
    ["main_category", "subcategory", "indicator_name"]
)

In [22]:
# Save to Excel
n_indicator_catalog.to_excel(
    "../outputs/n_indicator_catalog.xlsx",
    index=False
)

### Observations

To simplify exploration of the large dataset, indicator information for Palacio was selected as a representative neighbourhood. The main_category, subcategory, indicator_name and unit columns were exported to a catalogue for further review and feature selection.

### Selecting features for the ML model
The indicator catalogue was reviewed and 15 socioeconomic features were selected for the machine learning model. The catalogue was annotated during the review process and saved in the outputs folder as n_indicator_catalog_feature_selection.xlsx

#### Selected Features:

Some features required multiple indicators to be combined to calculate a percentage and are marked with an asterisk (*).

No.	Selected Feature Variable
- 1	Population density (inhabitants per hectare)
- 2	Percentage of population aged 25+ with higher education (university degree, postgraduate or doctorate) or diploma/technical degree *
- 3	Percentage of registered housing over total registered premises *
- 4	Average disposable income per person
- 5	Average annual net household income (Urban Audit)
- 6	Registered unemployment rate (February)
- 7	Population growth rate (%)
- 8	Average household size
- 9	Average age
- 10 Ageing index (% aged 65+)
- 11 Social welfare and equality vulnerability index
- 12 Economic and employment vulnerability index
- 13 Education and culture vulnerability index
- 14 Urban environment and mobility vulnerability index
- 15 Health vulnerability index

These selected features, and the component indicators required to create them, will be extracted and prepared during the data preparation stage